communications/serializers.py

```diff
@@ -0,0 +1,115 @@
+    """Serializer for the UserNotificationTypePreference model."""
+    notification_type = NotificationTypeSerializer(read_only=True)
+    notification_type_id = serializers.PrimaryKeyRelatedField(
+        queryset=NotificationType.objects.filter(is_active=True),
+        source='notification_type',
+        write_only=True
+    )
+    frequency = NotificationFrequencyField()
+    
+    class Meta:
+        model = UserNotificationTypePreference
+        fields = [
+            'id', 'notification_type', 'notification_type_id', 
+            'frequency', 'created_at', 'updated_at'
+        ]
+        read_only_fields = ['id', 'notification_type', 'created_at', 'updated_at']
+
+
+class UserNotificationPreferenceSerializer(serializers.ModelSerializer):
+    """Serializer for the UserNotificationPreference model."""
+    type_preferences = UserNotificationTypePreferenceSerializer(
+        source='type_preferences.all',
+        many=True,
+        read_only=True
+    )
+    user_id = serializers.PrimaryKeyRelatedField(
+        source='user',
+        read_only=True
+    )
+    
+    class Meta:
+        model = UserNotificationPreference
+        fields = [
+            'user_id', 'enable_in_app', 'enable_email', 'enable_push',
+            'type_preferences', 'created_at', 'updated_at'
+        ]
+        read_only_fields = ['user_id', 'created_at', 'updated_at']
+
+    def update(self, instance, validated_data):
+        # Update the main preference fields
+        instance.enable_in_app = validated_data.get('enable_in_app', instance.enable_in_app)
+        instance.enable_email = validated_data.get('enable_email', instance.enable_email)
+        instance.enable_push = validated_data.get('enable_push', instance.enable_push)
+        instance.save()
+        
+        # Handle type preferences if provided
+        type_prefs_data = self.initial_data.get('type_preferences', [])
+        for type_pref_data in type_prefs_data:
+            type_pref, _ = UserNotificationTypePreference.objects.get_or_create(
+                user_preference=instance,
+                notification_type_id=type_pref_data.get('notification_type_id')
+            )
+            type_pref.frequency = type_pref_data.get('frequency', type_pref.frequency)
+            type_pref.save()
+            
+        return instance
+
+
+class NotificationSerializer(serializers.ModelSerializer):
+    """Serializer for notifications."""
+    notification_type = NotificationTypeSerializer(read_only=True)
+    priority_display = serializers.CharField(
+        source='get_priority_display',
+        read_only=True
+    )
+    
+    class Meta:
+        model = Notification
+        fields = [
+            'notification_id',
+            'notification_type',
+            'title',
+            'message',
+            'is_read',
+            'priority',
+            'priority_display',
+            'created_at',
+            'expires_at',
+            'action_link',
```

Comment:
action_link is included in fields but is not defined in the serializer or the Notification model (at least in the provided code). This will raise an error unless action_link is a property or method on the model.